Adding Julia and Rust install instructions (#400)
EnzymeAD/Enzyme: content/Installation/Rust.md

---
title: "Rust"
date: 2019-11-29T15:26:15Z
draft: false
weight: 30
---

Getting started in Rust is easy, if you are using Linux. Otherwise, you probably need to wait for our next implementation, which will support Windows and Mac too.  
We have published `enzyme` on [crates.io](https://crates.io/crates/enzyme).

So after getting the following dependencies:

- git
- ninja
- cmake
- libssl-dev
- libclang-dev
- Rust (rustup) and any nightly toolchain 
- ~10GB free storage in $HOME/.cache

you can run
```sh
cargo install enzyme
```

This should only take a minute and give you an enzyme-install binary which you can use to build rustc/llvm/enzyme and a bunch of extra things, which might come helpful.  Now you can start the build process by calling
```bash
enzyme-install
```
It will build everything inside of `$HOME/.cache`, so please have around 10GB available there.

> Also take care, the build process takes around one hour on a recent 8 core cpu.

Reviewer: If I already have a `LLVMEnzyme-13.so` built, is there a way to use it from Rust without this time-consuming step? Besides compilation cost, I'm interested in using Enzyme from C and Rust within the same application.

Reviewer 2: @ZuseZ4, if I recall the issue here is not compiling Enzyme, but compiling Rust's LLVM, yes?

Also note that these instructions are aimed at the current work in progress nicer language integration (e.g. fewer compiler flag hacks + syntactic sugar) of Enzyme within Rust (see https://github.com/rust-ml/oxide-enzyme).

Using an existing Enzyme build (for a rust-compatible LLVM version) presumably would still work by either manually using the LLVM output of rust, or say using LLD as here (https://internals.rust-lang.org/t/automatic-differentiation-differential-programming-via-llvm/13188/7)

Author: Fair question. I was already wondering if I should add a note about just that while writing. The short version is maybe. 

As mentioned by @wsmoses, the main reason of having the current build procedure is stability. Each Rustc is only guaranteed to be compile-able against one specific LLVM commit (not even an LLVM release). So we could invest time to speed up the getting started process, but at the cost of introducing more bugs due to incompatibilities. Also, our main goal is a proper rustc integration, which is required due to some other issues. Once that's achieved we will archive the entire enzyme_build archive, so all optimizations here are only of limited duration. If you care about using C and rust together, 
I would just recommend you to use the clang version which I build along with rustc. This way Enzyme, Clang, rustc are
all based on the same LLVM version, which is probably the most reliable setup that you can get. That's what I'm using right now for some C++ code too. You will find clang in `/home/$USER/.cache/enzyme/rustc-1.57.0-src/build/x86_64-unknown-linux-gnu/llvm/bin/clang`. It might cost you two cpu-hrs, but it will probably safe you more hours of debugging/developing.

However, if you prefer to test your existing Enzyme version together with an official rustc, here is some background /history:
I added a flag to rustc, which allows loading LLVM plugins. This flag will only be available if you build rustc from source, as it is done by my script. Having this plugin interface has two purposes:
1) we can run some specific Enzyme optimizations before AD
2) We could use the __enzyme_ notation, as it is done in the C++ examples.

However: we moved away from the __enzyme notation and we haven't added the usage of the enzyme optimizations yet,
so you should theoretically be able to use some other rustc. You can do that by replacing `+enzyme` with `+nightly`
when building a crate. I just tested it an it fails due to having `__rust_probestack` defined more than once. I handle this [here](https://github.com/rust-ml/oxide-enzyme/blob/68a1f1eec636eeb1a07f273660abd72a9d5f7eba/src/lib.rs#L548), 
so you could try to debug why it's working with the enzyme, but not with the nightly toolchain.
Finally you could try to build rustc based on your existing LLVM Version, but as mentioned, rustc devs don't offer any
guarantees about that working. Also, building rustc-stage2 even without llvm takes a descent amount of time.

So yes, you might be able to sip this time-consuming step, I just really don't know if it's recommendable atm.
Either way, just let me know what you prefer, and I can try to help with that.

Reviewer: Thanks for all these details. I'll just follow the procedure for now, though I hope it'll be possible to upstream what is necessary.

I'd be okay using one build of Enzyme with release clang for differentiating C and another built internally for rustc. Enzyme builds pretty fast and distribution will be a hassle if one has to always match. (For example, users may build our libraries on an HPC machine with vendor MPI set up to use their version of clang. Building rustc+llvm from source and then using environment variables to make the vendor MPI use our special clang will be an error-prone user experience.)

One follow-up: what's the process for incremental rebuilds instead of the released crate, say upon pulling Enzyme or if one is hacking on Enzyme?

Author: So I just pushed a new version to crates.io yesterday, which moved to the latest enzyme release 0.0.25. I've added some minimal logic to recognize whether enzyme or rustc are outdated, and only build what is needed.

If you made some changes to enzyme locally and want to test them, I recommend to just build it against the llvm from the rustc-1.57 build.
If you place your enzyme repo folder inside of '$HOME/.cache/enzyme' you can then update the ENZYME_VER in oxide-enzyme/src/lib.rs to whatever name you used for your folder.
You also need to adjust the LD_LIBRARY_PATH, but then you should be ready. I will try to add this also to the README before publishing the next enzyme version on crates.io.

## Verifying installation

Calling `enzyme-install` will automatically run Enzyme's unit tests after finishing the build process, so you should see something like
```sh
Testing Time: 0.63s
Passed : 299
Expectedly Failed: 5
```

You can also run a few basic Rust tests:
```sh
git clone https://github.com/rust-ml/oxide-enzyme
cd oxide-enzyme/example
cargo enzyme
```
You will find your binary slightly more hidden than usual in `target/$TARGET/debug/example`.

## Using Enzyme

> In order to play nicely with cargo, the Rust workflow differs somewhat from the [C++](https://enzyme.mit.edu/getting_started/UsingEnzyme/) using Enzyme section.

Reviewer 2: Change C++ to generic?

Author: So, you still think we should merge this (in an updated manner)?
I thought we would just leave the installation instruction as they are and just add the links to the julia and rust website, as mentioned above?

Reviewer 2: At minimum right now more docs are better than none, so having rust instructions is definitely a plus.

I do think the more elegant solution is that this part of the Rust documentation is at enzyme.mit.edu/rust, which we can still move to after the fact.

I'd say for now here this would still be useful as a way to better organize the installation page into a top level (link to from source / homebrew package manager / docker container / rust+julia package manager sections), and corresponding sections with docs for.

Reviewer 2: If it's okay with you, do you mind pinging me again when some of the changes have been made for another look before merging?

In particular, the generic instructions are used by folks who play with other languages [like Swift/Fortran], and I don't want them getting lost/confused when there's no section for their language.

> First, you need to specify in your `build.rs` file how which function should be differentiated. 

Reviewer: ```suggestion
First, you need to specify in your `build.rs` file how each function should be differentiated. 
```
would be useful to link to an example. Also, is this `build.rs` approach a short-term solution or is it something you envision supporting longer term? I was hoping for a macro that could go in the source, ideally either in the module with the definition or in another module.

Afterwards, you can run `cargo enzyme` without additional arguments, to build your crate.

> If you want to have some more controll, you can also modify the following build command, which `cargo enzyme` is using:

Reviewer: ```suggestion
If you want to have some more control, you can also modify the following build command that `cargo enzyme` uses:
```

```sh
RUSTFLAGS="--emit=llvm-bc" cargo +enzyme -Z build-std rustc --target x86_64-unknown-linux-gnu -- --emit=llvm-bc -g -C opt-level=3 -Zno-link && RUSTFLAGS="--emit=llvm-bc" cargo +enzyme -Z build-std rustc --target x86_64-unknown-linux-gnu -- --emit=llvm-bc -g -C opt-level=3
```

Please be aware that Enzyme currently can't be used in dependencies. However you can use Enzyme to differentiate functions and types defined in your dependencies. 
We still have quite a few issues open and are lacking some documentation. So you might find [this issue](https://github.com/rust-ml/oxide-enzyme/issues/6) and 

> [this C++ documentation](https://enzyme.mit.edu/getting_started/CallingConvention/#types) useful to understand the Rust frontend. 

Reviewer 2: Change "this C++ documentation" to "Enzyme argument convention documentation"?

> If you have specific question please open an issue and if you have some time, we are always happy to review PR's which add documentation or examples.

Reviewer 2: Can you add a `for developers` section which describes how one can use the Rust frontend with a custom build/commit of Enzyme (for the purpose of development/debugging). 

See https://enzyme.mit.edu/julia/dev_docs/ for a Julia equivalent.
